File: web/src/features/annotation/types/annotation.ts
export type SelectorType =
  | "FULL_ASSET"
  | "RECORD"
  | "TEXT_SPAN"
  | "TIME_RANGE"
  | "FRAME_RANGE"
  | "DOCUMENT_PAGE";

export type ResultType =
  "bbox" | "polygon" | "text_region" | "caption" | "classification" | "ner";

export type RevisionSource = "human" | "machine";

export interface AnnotationResult {
  id?: string;
  output_id?: string;
  category_id?: string | null;
  result_type?: string;
  value?: unknown;
  geometry?: {
    x?: number;
    y?: number;
    width?: number;
    height?: number;
    points?: number[][];
    [key: string]: unknown;
  } | null;
  payload?: Record<string, unknown> | null;
  attributes?: Record<string, unknown> | null;
  confidence?: number;
  created_at?: string;
}

export interface AnnotationRevision {
  id: string;
  annotation_id: string;
  revision_number: number;
  ontology_version_id: string;
  created_by: string;
  source: RevisionSource;
  created_at?: string;
  results: AnnotationResult[];
  category_ids?: string[];
}

export interface Annotation {
  id: string;
  asset_id: string;
  project_id: string;
  target_type: SelectorType;
  target_selector: Record<string, unknown>;
  ontology_version_id?: string;
  created_by: string;
  created_at?: string;
  updated_at?: string;
  latest_revision?: AnnotationRevision;
  revisions?: AnnotationRevision[];
}

export interface CreateAnnotationRequest {
  asset_id: string;
  project_id: string;
  ontology_version_id: string;
  target_type?: SelectorType;
  target_selector?: Record<string, unknown>;
  source?: RevisionSource;
  results: AnnotationResult[];
}

export interface CreateRevisionRequest {
  ontology_version_id: string;
  source?: RevisionSource;
  results: AnnotationResult[];
}

File: packages/domain/domain/entities/annotation.py
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Literal

ResultType = Literal[
    "bbox", "polygon", "text_region", "caption", "classification", "ner"
]
RevisionSource = Literal["human", "machine"]


@dataclass
class BBoxGeometry:
    x: float
    y: float
    width: float
    height: float


@dataclass
class PolygonGeometry:
    points: list[list[float]]  # [[x1, y1], [x2, y2], ...]


@dataclass
class AnnotationResultEntity:
    id: str
    revision_id: str
    result_type: ResultType
    category_id: str | None = None
    geometry: dict[str, Any] | None = None
    payload: dict[str, Any] | None = None
    attributes: dict[str, Any] | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


@dataclass
class AnnotationRevisionEntity:
    id: str
    annotation_id: str
    revision_number: int
    created_by: str
    source: RevisionSource = "human"
    created_at: datetime | None = None
    updated_at: datetime | None = None
    results: list[AnnotationResultEntity] = field(default_factory=list)


@dataclass
class AnnotationEntity:
    id: str
    asset_id: str
    project_id: str
    ontology_version_id: str
    created_by: str
    label_studio_task_id: int | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    revisions: list[AnnotationRevisionEntity] = field(default_factory=list)
